include/aws/crt/mqtt/private/MqttConnectionCore.h

```diff
@@ -0,0 +1,383 @@
+#pragma once
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/crt/Exports.h>
+#include <aws/crt/StlAllocator.h>
+#include <aws/crt/Types.h>
+#include <aws/crt/http/HttpConnection.h>
+#include <aws/crt/io/SocketOptions.h>
+#include <aws/crt/io/TlsOptions.h>
+#include <aws/crt/mqtt/MqttTypes.h>
+
+#include <aws/mqtt/client.h>
+#include <aws/mqtt/v5/mqtt5_client.h>
+
+#include <functional>
+#include <memory>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        namespace Mqtt
+        {
+            class MqttConnection;
+
+            /**
+             * @internal
+             * The MqttConnectionCore is an internal class for MqttConnection. The class is used to handle communication
+             * between MqttConnection and underlying C MQTT client. This class should only be used internally by
+             * MqttConnection.
+             * @note Unless otherwise specified, all function arguments need only to live through the duration of the
+             * function call.
+             */
+            class AWS_CRT_CPP_API MqttConnectionCore final : public std::enable_shared_from_this<MqttConnectionCore>
+            {
+              public:
+                /**
```

Comment:
The intention was to get rid of the private constructors which make impossible to use `make_shared` and force adding friends for the whole class.
If you think it's overcomplicated, it's really easy to revert back to private constructors.

Reply:
Feels a bit weird.  Given the existing pattern, let's just stick with it since there doesn't seem to be a clear benefit.  In general, we can definitely break with past patterns when improvements are clear, but this leaves me in kind of a maybe?/maybe not? state of mind.